### PR TITLE
[otbn,doc] Add README for RIG internals

### PR DESCRIPTION
**Pull request by ctopal**
I am also planning to add an example of changing the internals to hit a coverage hole but I thought better to have it here for the first feedback.

### CODE REVIEW

**Review by cindychip — COMMENTED** [review]
not super familiar with OTBN RIG, but from the description, it LGTM ;)

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+This main script also constructs a `Program` and a `Model`.
+`Program` reflects both data memory that gets initialised and also the instruction memory that gets set with the generated snippets.
+
+Generated program then gets used in `otbn-rig` to create proper output files to simulate the generated instructions.
+
+## Snippet Generators
+These generators are split up into two lists.
+The list called `_cont_generators` is the list of snippet generators that will not produce snippets that ends the program.
+The other list `_end_generators` consists of generators that produces either an `ECALL` or a fault to finish the program.
+The type of fault that the snippet generator creates is in its name (e.g. BadInsn snippet generator will generate snippets that end with an illegal instruction).
```
Comment (marnovandermaas) [inline hw/ip/otbn/dv/rig/rig/README.md:20]:
```suggestion
The type of fault that the snippet generator creates is in its name (e.g. `BadInsn` snippet generator will generate snippets that end with an illegal instruction).
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+### Configuring a Snippet Generator
+Each snippet generator has its own weight specified in `dv/rig/rig/configs` folder.
+In it's `README`, a detailed explanation about how to configure specific `.yaml` files for wanted behavior exists.
```
Comment (marnovandermaas) [inline hw/ip/otbn/dv/rig/rig/README.md:24]:
```suggestion
In its `README`, there is a detailed explanation about how to configure specific `.yaml` files for wanted behavior.
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+- `loopy.yaml`: A custom configuration that generates hundred times as many loop instructions compared to the `base` configuration.
+- `straigh-line.yaml`: A custom configuration that does not generate any branches, hence the name.
+
+## Abstract Models in RIG
+In order to generate concise programs that don't run into architectural problems RIG needs to have an idea about how OTBN works, what's a memory, a register etc.
+This is all handled in `model.py` in which different classes for different models exists.
+A list of them includes:
+- `CallStack`: Abstract model of the `x1` call stack
+- `LoopStack`: An abstract model of the loop stack.
+It has a `stuck` counter for keeping track of pop addresses that doesn't match -related with "Loop Warping".
```
Comment (marnovandermaas) [inline hw/ip/otbn/dv/rig/rig/README.md:39]:
```suggestion
It has a `stuck` counter for keeping track of pop addresses that doesn't match - related with "Loop Warping".
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+- `Model`: Abstract model of the processor and the memories.
+Holds tracks of registers and memory locations that have defined values at a current instruction stream.
+Note that while in operation there can be more than two models instantiated to enable RIG to make decisions based on the current state of the program, and to maintain the consistency of the program state.
+For more details see `loop.py`
```
Comment (marnovandermaas) [inline hw/ip/otbn/dv/rig/rig/README.md:43]:
```suggestion
For more details see `loop.py`.
```

**Review by marnovandermaas — COMMENTED** [review]
LGTM, just a few small suggestions.

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# OTBN Random Instruction Generator - Internals
+
+This directory contains the internal part of the random instruction generator (RIG).
+This document aims to help with the general understanding of the whole script.
+
+## Overall Look
+`rig.py` calls `snippet_gens.py` to generate a configured list of smaller snippet generators.
+The main constraint while generating these snippets is called `fuel` and it's an upper bound for the number of instructions to execute.
+The SnippetGens class invloves a list of these smaller snippet generators defined in `_CLASSES`.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/rig/rig/README.md:9]:
```suggestion
The `SnippetGens` class involves a list of these smaller snippet generators defined in `_CLASSES`.
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+This main script also constructs a `Program` and a `Model`.
+`Program` reflects both data memory that gets initialised and also the instruction memory that gets set with the generated snippets.
+
+Generated program then gets used in `otbn-rig` to create proper output files to simulate the generated instructions.
+
+## Snippet Generators
+These generators are split up into two lists.
+The list called `_cont_generators` is the list of snippet generators that will not produce snippets that ends the program.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/rig/rig/README.md:18]:
```suggestion
The list called `_cont_generators` is the list of snippet generators that will not produce snippets that end the program.
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# OTBN Random Instruction Generator - Internals
+
+This directory contains the internal part of the random instruction generator (RIG).
+This document aims to help with the general understanding of the whole script.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/rig/rig/README.md:4]:
```suggestion
This document aims to help with the general understanding of the RIG.
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+## Overall Look
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/rig/rig/README.md:6]:
```suggestion
## Overall Structure
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+`rig.py` calls `snippet_gens.py` to generate a configured list of smaller snippet generators.
+The main constraint while generating these snippets is called `fuel` and it's an upper bound for the number of instructions to execute.
+The SnippetGens class invloves a list of these smaller snippet generators defined in `_CLASSES`.
+
+This main script also constructs a `Program` and a `Model`.
+`Program` reflects both data memory that gets initialised and also the instruction memory that gets set with the generated snippets.
+
+Generated program then gets used in `otbn-rig` to create proper output files to simulate the generated instructions.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/rig/rig/README.md:14]:
```suggestion
The generated program then gets used in `otbn-rig` (which lives in the parent directory) to create proper output files to simulate the generated instructions.
```

File: hw/ip/otbn/dv/rig/rig/README.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+## Snippet Generators
+These generators are split up into two lists.
+The list called `_cont_generators` is the list of snippet generators that will not produce snippets that ends the program.
+The other list `_end_generators` consists of generators that produces either an `ECALL` or a fault to finish the program.
+The type of fault that the snippet generator creates is in its name (e.g. BadInsn snippet generator will generate snippets that end with an illegal instruction).
+
+### Configuring a Snippet Generator
+Each snippet generator has its own weight specified in `dv/rig/rig/configs` folder.
```
Comment (andreaskurth) [inline hw/ip/otbn/dv/rig/rig/README.md:23]:
```suggestion
Each snippet generator has its own weight specified in the `configs` folder.
```

**Review by andreaskurth — COMMENTED** [review]
Thanks for writing this up @ctopal! I have a couple of suggestions to clarify points I did not understand right away :slightly_smiling_face:

**Comment by ctopal** [comment]
Applied the feedback, thanks for having a look everyone!

**Review by marnovandermaas — APPROVED** [review]
LGTM

**Review by andreaskurth — APPROVED** [review]
Nice, thanks @ctopal!

**Pull request merged**